// web3-custom/index.js
// extending web3 support to admin and other methods via rpc/ipc web3 api.
// based on https://github.com/ethereum/go-ethereum/commit/5f8e5a487588bab49fb3c9321fd8903132426c47?diff=split

var web3 = require('web3');


// admin methods from 
//https://github.com/ethereum/go-ethereum/blob/5f8e5a487588bab49fb3c9321fd8903132426c47/rpc/api/admin_js.go
web3._extend({
	property: 'admin',
	methods:
	[
		new web3._extend.Method({
			name: 'addPeer',
			call: 'admin_addPeer',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'peers',
			call: 'admin_peers',
			params: 0,
			inputFormatter: [],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'exportChain',
			call: 'admin_exportChain',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'importChain',
			call: 'admin_importChain',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'verbosity',
			call: 'admin_verbosity',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'syncStatus',
			call: 'admin_syncStatus',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputInt],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'setSolc',
			call: 'admin_setSolc',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputString
		})
	],
	properties:
	[
		new web3._extend.Property({
			name: 'nodeInfo',
			getter: 'admin_nodeInfo',
			outputFormatter: web3._extend.formatters.formatOutputString
		})
	]
});

//debug
//https://github.com/ethereum/go-ethereum/blob/5f8e5a487588bab49fb3c9321fd8903132426c47/rpc/api/debug_js.go

web3._extend({
	property: 'debug',
	methods:
	[
		new web3._extend.Method({
			name: 'printBlock',
			call: 'debug_printBlock',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputString
		}),
		new web3._extend.Method({
			name: 'getBlockRlp',
			call: 'debug_getBlockRlp',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputString
		}),
		new web3._extend.Method({
			name: 'setHead',
			call: 'debug_setHead',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'processBlock',
			call: 'debug_processBlock',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'seedHash',
			call: 'debug_seedHash',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputString
		})
	],
	properties:
	[
	]
});


//miner
//from https://github.com/ethereum/go-ethereum/blob/5f8e5a487588bab49fb3c9321fd8903132426c47/rpc/api/miner_js.go
web3._extend({
	property: 'miner',
	methods:
	[
		new web3._extend.Method({
			name: 'start',
			call: 'miner_start',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'stop',
			call: 'miner_stop',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'getHashrate',
			call: 'miner_hashrate',
			params: 0,
			inputFormatter: [],
			outputFormatter: web3._extend.utils.toDecimal
		}),
		new web3._extend.Method({
			name: 'setExtra',
			call: 'miner_setExtra',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'setGasPrice',
			call: 'miner_setGasPrice',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'startAutoDAG',
			call: 'miner_startAutoDAG',
			params: 0,
			inputFormatter: [],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'stopAutoDAG',
			call: 'miner_stopAutoDAG',
			params: 0,
			inputFormatter: [],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'makeDAG',
			call: 'miner_makeDAG',
			params: 1,
			inputFormatter: [web3._extend.formatters.inputDefaultBlockNumberFormatter],
			outputFormatter: web3._extend.formatters.formatOutputBool
		})
	],
	properties:
	[
		new web3._extend.Property({
			name: 'hashrate',
			getter: 'miner_hashrate',
			outputFormatter: web3._extend.utils.toDecimal
		})
	]
});

//network
//from https://github.com/ethereum/go-ethereum/blob/5f8e5a487588bab49fb3c9321fd8903132426c47/rpc/api/net_js.go
web3._extend({
	property: 'network',
	methods:
	[
		new web3._extend.Method({
			name: 'addPeer',
			call: 'net_addPeer',
			params: 1,
			inputFormatter: [web3._extend.utils.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Method({
			name: 'id',
			call: 'net_id',
			params: 0,
			inputFormatter: [],
			outputFormatter: web3._extend.formatters.formatOutputString
		}),
		new web3._extend.Method({
			name: 'getPeerCount',
			call: 'net_peerCount',
			params: 0,
			inputFormatter: [],
			outputFormatter: web3._extend.formatters.formatOutputString
		}),
		new web3._extend.Method({
			name: 'peers',
			call: 'net_peers',
			params: 0,
			inputFormatter: [],
			outputFormatter: function(obj) { return obj; }
		})
	],
	properties:
	[
		new web3._extend.Property({
			name: 'listening',
			getter: 'net_listening',
			outputFormatter: web3._extend.formatters.formatOutputBool
		}),
		new web3._extend.Property({
			name: 'peerCount',
			getter: 'net_peerCount',
			outputFormatter: web3._extend.utils.toDecimal
		})
	]
});

//personal
// from https://github.com/ethereum/go-ethereum/blob/5f8e5a487588bab49fb3c9321fd8903132426c47/rpc/api/personal_js.go
web3._extend({
	property: 'personal',
	methods:
	[
		new web3._extend.Method({
			name: 'listAccounts',
			call: 'personal_listAccounts',
			params: 0,
			inputFormatter: [],
			outputFormatter: function(obj) { return obj; }
		}),
		new web3._extend.Method({
			name: 'newAccount',
			call: 'personal_newAccount',
			params: 1,
			inputFormatter: [web3._extend.formatters.formatInputString],
			outputFormatter: web3._extend.formatters.formatOutputString
		}),
		new web3._extend.Method({
			name: 'unlockAccount',
			call: 'personal_unlockAccount',
			params: 3,
			inputFormatter: [web3._extend.formatters.formatInputString,web3._extend.formatters.formatInputString,web3._extend.formatters.formatInputInt],
			outputFormatter: web3._extend.formatters.formatOutputBool
		})
	],
	properties:
	[
	]
});


module.exports = {
  web3:web3,
}
